src/main/java/tdl/client/queue/QueueBasedImplementationRunner.java

```diff
@@ -44,78 +220,214 @@ public Builder withSolutionFor(String methodName, UserImplementation userImpleme
+        try {
+            requests.clear();
+            requests.addAll(receive());
+
             audit.logLine("Waiting for requests");
-            Optional<Request> request = remoteBroker.receive();
-            while (request.isPresent()) {
-                request = applyProcessingRules(request.get(), deployProcessingRules, remoteBroker);
+            logToConsole("        QueueBasedImplementationRunner requests: " + requests.size());
+
+            for (int requestIndex = requests.size() - 1; requestIndex >= 0; requestIndex--) {
+                Request request = requests.get(requestIndex);
+                logToConsole("        QueueBasedImplementationRunner applyProcessingRules [start]");
+                audit.startLine();
+                audit.log(request);
+                requests.add(request);
+
+                //Obtain response from user
+                Response response = deployProcessingRules.getResponseFor(request);
+                audit.log(response);
+
+                //Act
+                if (response instanceof FatalErrorResponse) {
+                    audit.endLine();
+                } else {
+                    respondTo(request, with(response));
+                    audit.endLine();
+
+                    logToConsole("        QueueBasedImplementationRunner deleting consumed message");
+                    deleteMessage(request.getOriginalMessage());
+                }
             }
         } catch (Exception e) {
+            logToConsole("        QueueBasedImplementationRunner run [error]");
             String message = "There was a problem processing messages";
             LOGGER.error(message, e);
             audit.logException(message, e);
         }
         audit.logLine("Stopping client");
+        logToConsole("        QueueBasedImplementationRunner run [end]");
+    }
+
+    <T> T with(T obj) {
+        return obj;
     }
 
     public int getRequestTimeoutMillis() {
         return config.getRequestTimeoutMillis();
     }
 
-    private Optional<Request> applyProcessingRules(
-            Request request, ProcessingRules processingRules, RemoteBroker remoteBroker)
-            throws BrokerCommunicationException {
-        audit.startLine();
-        audit.log(request);
+    private List<Request> receive() throws BrokerCommunicationException {
+        logToConsole("     QueueBasedImplementationRunner receive [start]");
+        try {
+            List<Message> messages;
+            List<Request> successfulMessages = new ArrayList<>();
+            boolean continueFetching;
+            do {
+                messages = client.receiveMessage(receiveRequestQueueMessage).getMessages();
+                continueFetching = messages.size() > 0;
+                if (continueFetching) {
+                    logToConsole("     QueueBasedImplementationRunner messages: " + messages);
+                    List<Request> newResult = new ArrayList<>();
+                    for (Message message : messages) {
```

Comment:
The translation from a Message object to a Request object can be encapsulated, I am confident that this is a good abstraction.

Reply:
Can we mark this as a TODO, since we are still in spiking mode